centrallix/expression/exp_functions.c

```diff
@@ -4159,6 +4161,127 @@ int exp_fn_cos_compare(pExpression tree, pParamObjects objlist, pExpression i0,
     return 0;
     }
 
+
+/*
+ * exp_fn_argon2id
+ * This method hashes a given password using the Argon2 algorithm (ID variant)
+ *
+ * Parameters:
+ * 	pExpression tree: 
+ * 	pParamObjects: 
+ * 	pExpression passowrd: The password, passed as a pExpression
+ * 	pExpression salt: The salt, passed as a pExpression
+ *
+ * returns:
+ *	 0 if successful
+ *	 -1 if error
+ */
+int exp_fn_argon2id(pExpression tree, pParamObjects objlist, pExpression password, pExpression salt)
+{
+
+    //check password and salt
+    if (!password || !salt)
+	{
+	mssError(1, "EXP", "Invalid Parameters: Method requires two arguments");
+	return -1;
+	}
+    else if ((password->Flags | salt->Flags) & EXPR_F_NULL)
+	{
+	tree->Flags |= EXPR_F_NULL;
+	tree->DataType = DATA_T_STRING;
+	return 0;
+	}
+
+    // The default values of the following four variables should be tuned for each specific system's needs
+    // T_COST determines the number of passes the algorithm makes
+    unsigned int T_COST = 2; 
+    if ((tree->Children.nItems >= 3) && 
+	((tree->Children.Items[2]) != NULL) && 
+	((pExpression)tree->Children.Items[2])->DataType == DATA_T_INTEGER && 
+	((pExpression)tree->Children.Items[2])->Integer < 8 && 
```

Comment:
Is 8 the number of iterations?  If so I'd guess this would need to be much higher, so we can cause the algorithm to take up to about a second to run.

Reply:
The upper bound for T_COST is now set to 24. I consistently got approximately a 1 second execution time with 12 iterations.